Railsでテストを書く

naokinaokiboo opened the pull request:
「Railsでテストを書く」の課題内容をpushしました。
レビュー、よろしくお願い致します。🙇‍♂️
※テストデータの作成には、FactoryBotを利用しました。

naokinaokiboo committed:
fixturesファイルを削除（テストデータの作成はfactory botを使用するため不要）

naokinaokiboo committed:
コントローラー用のテストファイルを削除

naokinaokiboo committed:
factory_bot_rails gem を追加

naokinaokiboo committed:
factory bot の使用にあたっての設定を追加

・メソッド呼び出しにクラス名(FactoryBot)を省略するための設定

naokinaokiboo committed:
Userクラスのモデルテストを追加

naokinaokiboo committed:
Reportクラスのモデルテストを追加

naokinaokiboo committed:
不要なシステムテストファイルを削除

naokinaokiboo committed:
日報のCRUD処理に対してのシステムテストを追加

yoshitsugu commented on test/models/report_test.rb:
```diff
@@ -3,7 +3,61 @@
+    FactoryBot.rewind_sequences
+  end
+
+  test 'editable? should return true' do
```
ここだけ読むと `editable?` は常にtrueを返すのかな？と思ってしまうので、「when ~」など条件をつけたほうがよいかと思います。
また、メソッドの単体テストは単にテスト対象のメソッド名になることも多いです。

```ruby
test '#editable?' do
  report = build(:report_with_user)
  assert_equal true, report.editable?(report.user)
  report2 = build(:report_with_user)
  assert_equal false, report2.editable?(report1.user)
end
```
のような感じですね。

naokinaokiboo replied:
確かに、説明文が中途半端でした。💦
メソッドの単体テストの場合、以下のいずれかになるように注意するようにします。
・説明を記載するのであれば、正確に記載する
・もしくは、テスト対象のメソッド名にする

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -4,44 +4,76 @@
+    fill_in '内容', with: 'テストデータの作成にfixturesを使用した後、FactoryBotも試してみた'
+    click_on '登録する'
+
+    assert_text '日報が作成されました。'
```
作成された日報が表示されることまで確認しましょう。

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -4,44 +4,76 @@
+    fill_in '内容', with: 'テストデータの作成にfixturesを使用した後、FactoryBotも試してみた'
+    click_on '更新する'
+
+    assert_text '日報が更新されました。'
```
更新された日報が表示されることまで確認しましょう。

yoshitsugu commented on test/system/reports_test.rb:
```diff
@@ -4,44 +4,76 @@
+    click_on 'この日報を削除'
 
-    assert_text 'Report was successfully destroyed'
+    assert_text '日報が削除されました。'
```
削除された日報が表示されなくなったことまで確認しましょう。

yoshitsugu reviewed (commented):
コメントしました！

naokinaokiboo committed:
Reportクラスの単体テストの説明文を修正

naokinaokiboo committed:
日報の作成、更新、削除のシステムテストの確認内容を追加

・追加した確認内容
 - 日報の作成後に、その日報が表示されていること
 - 日報の更新前に表示されていないタイトルと内容が、更新後に表示されていること
 - 日報の削除前に表示されているタイトルと内容がが、削除後に表示されないこと

naokinaokiboo commented:
ご指摘頂いた点について、修正をpush致しました。
再度、ご確認よろしくお願い致します。🙇‍♂️

yoshitsugu reviewed (approved):
OKです！
